Qutrit mixed measure

Gabriel-Bottrill opened the pull request:
Context:
Currently the qutrit_mixed device is being developed this is a necessary addition to the overall project as this measures a qutrit mixed state. This is a prerequisite for other functionality relating to the qutrit mixed device for noisy qutrit simulation.

Description of the Change:
Added functionality for measuring a qutrit mixed-state. The new measure function can be used to measure a qutrit mixed-state.

Benefits:
Allows for measurements of a mixed state, will be used to add execute functionality to qutrit mixed-state device allowing for noisy qutrit simulation

Possible Drawbacks:
Abstracting for qubits and more generally qutrits may have added challenges and will require a refactor or copied code code-smell. Methods are not the fastest possible, optimizations can still be made

Related GitHub Issues:
N/A

Gabriel-Bottrill commented:
This is still fairly early, I meant to have it closer to finishing by now but I got behind with apply_operations. I just wanted to put it up so you all would know it's being worked on and will be finished soon! @albi3ro @glassnotes @mudit2812 @trbromley .

Gabriel-Bottrill commented:
measurementprocess.diagonalizing_gates() is throwing errors due to the observables being for qutrits so I added methods to compute some of the main measurable values. I can leave it like this, but it slightly defeats the purpose of the new device structure. Any suggestions on getting around this would be very helpful, thanks!

Gabriel-Bottrill commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,308 @@
+    return math.where(probs < 0, -probs, probs)
+
+
+def calculate_probability(
```
a similar method could be use to define ProbabilityMP calculate_dm

Gabriel-Bottrill commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,308 @@
+    return probs
+
+
+def calculate_variance(
```
essentially copied from VarianceMP calculate_state

Gabriel-Bottrill commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,308 @@
+    )
+
+
+def sum_of_terms_method(
```
copied from qubit measure, only difference is measure_func, could instead change the qubit version to have a measure_func. From discussions it seems ok to have copied code for now

albi3ro replied:
I'm curious about the use case for the `measure_func` ? Doesn't seem like it's getting used right now.

Gabriel-Bottrill replied:
It is not used as is. I was suggesting that if the qubit version was changed to include a `measure_func` then `sum_of_terms_method` could be entirely removed from this file. Instead, there could be a call to  `devices.qubit.measure.sum_of_terms_method` with `measure_func = qutrit_mixed.measure`. This abstraction is not consistent with the plans for this qutrit mixed state device as we discussed the qubit device functionality shouldn't be abstracted to work for qutrits. 

I will remove the `measure_func`.

Gabriel-Bottrill commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,308 @@
+    return math.real(trace)
+
+
+def reduce_density_matrix(  # TODO: ask if I should have state diagonalization gates?
```
I'm unsure if I should have state diagonalization gates being applied here because it seems weird to have a measurement of the state with an observable

glassnotes replied:
Yeah it would be equivalent to getting the state represented in a different basis, I guess. I don't think the regular devices even have support for that.

Gabriel-Bottrill reviewed (commented):
This branch is branched from apply state so it will have some extra  "bulk" until apply state is merged. Commented where new code is and commented on what is copied code that could be abstracted to measurements.

Gabriel-Bottrill commented:
@albi3ro, @glassnotes, @mudit2812, @trbromley I have the measure functionality here, I will have to remerge it once applystate is merged, but it is ready to be reviewed.

Gabriel-Bottrill commented:
@albi3ro, @glassnotes, @mudit2812, @trbromley now that apply_operation is done this PR is ready to be reviewed!

Gabriel-Bottrill commented:
Also should I squash the commits or is it better to have the commit history?

albi3ro commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,308 @@
+alphabet_array = math.asarray(list(alphabet))
+
+
+def apply_observable_einsum(obs: Observable, state, is_state_batched: bool = False):
```
How does this differ from `apply_operation_einsum`?

glassnotes replied:
Gabe can elaborate more but `apply_operation_einsum` handles batches of Kraus operators and thus operation on the state from both sides. I guess these could be merged/streamlined by redefining this one generically in `utils`, to just perform the matrix multiplication, and then using it in this module straight up (and adjusting the `apply_operation` one to apply it twice.

Gabriel-Bottrill replied:
Yes

1. `apply_operation_einsum` takes operation (with matrix $A$) and state with density martix $\rho$ and returns $A\rho A^\dagger$ and in the case of a channel it returns $\sum_i K_i\rho K_i^\dagger$. where K is the set of Kraus matrices
2. `apply_observable_einsum` takes observable (with matrix $O$) and state with density martix $\rho$ and returns $O\rho$ this is used to take the trace for expectation value $<O>=Tr(O\rho)$

I could definitly implement it as @glassnotes suggests, but I think it would actually complicate things and make `apply_operation` slower as it currently uses the equivalent of:
```math.einsum(f"...{kraus_indices},...{state_indices},...{kraus_dag_indices}->...{new_state_indices}", kraus, state, kraus_dagger)```
I think by separating this would require an equivalent amount of repeated code.

It may make more sense to instead require diagonalizing gates for all measurements and not include the trace method if this repetition is not ideal

albi3ro replied:
Thanks for the explanation 👍

glassnotes replied:
Makes sense, thanks @Gabriel-Bottrill !

glassnotes commented on doc/releases/changelog-dev.md:
```diff
@@ -135,6 +135,9 @@
 * A function called `apply_operation` has been added to the new `qutrit_mixed` module found in `qml.devices` that applies operations to device-compatible states.
   [(#5032)](https://github.com/PennyLaneAI/pennylane/pull/5032)
 
+* A function called `measure` has been added to the new `qutrit_mixed` module found in `qml.devices` that measures device-compatible states for a collection of measurement processes.
```
It would be more concise to merge all the relevant device additions into a single CHANGELOG entry with all the PRs listed after. Or is that against convention? (Gabe no need to update yet, we can see if anyone has any feedback)

albi3ro replied:
Sometimes we do so, see:

https://github.com/PennyLaneAI/pennylane/blob/d0c435d15ddcd9a28f7ff7104fceef6e4304ab42/doc/releases/changelog-dev.md?plain=1#L123

But we also clean up and reorganize the changelog before each release, at which point each point might get combined.

glassnotes replied:
Great, let's leave it for now then!

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,308 @@
+    def map_indices(
+        state_indices, row_indices, new_row_indices, **kwargs
+    ):  # pylint: disable=unused-argument
+        """map indices to wires
```
```suggestion
        """Map indices to wires
```

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,308 @@
+        TensorLike: the result of obs@state
+    """
+
+    def map_indices(
```
Similar comment, can something more general be defined for the mapping and then used for both the measurement and operation? (If not, to match the convention of the other file, this can be defined as outside the function).

Gabriel-Bottrill replied:
I think I may have over complicated this.
In `utils` there is a function called `get_einsum_mapping`:

```
def get_einsum_mapping(
    op, state, indices_mapping_function, is_state_batched):
   # find  indices of operator and state
   ...

  # apply mapping function
      return indices_mapping_function(
          state_indices=state_indices,
          kraus_index=kraus_index,
          row_indices=row_indices,
          new_row_indices=new_row_indices,
          col_indices=col_indices,
          new_col_indices=new_col_indices,
    )
```

I changed `map_indices` variable -> `indices_mapping_function` here for clarity.

`map_indices` here and `_map_indices_apply_channel` are inputted as `indices_mapping_function`'s value to find the indices mapping to put into einsum.

another option would be to make get_einsum_mapping to return a dictionary of the values and have `map_indices` here and `_map_indices_apply_channel`  call `get_einsum_mapping`  (or getting rid of these helpers entirely).

glassnotes replied:
Oh okay I see - I think for consistency then, this should be defined out of the scope of `apply_observable_einsum`, like is done with `_map_indices_apply_channel` in the `apply_operation.py` file (as @albi3ro suggests in the comment below)

Gabriel-Bottrill replied:
I agree, will add in next commit

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,308 @@
+    Args:
+        measurementprocess (ExpectationMP): measurement process to apply to the state
+        state (TensorLike): the state to measure
+        is_state_batched (bool): whether the state is batched or not
+    Returns:
```
```suggestion
        is_state_batched (bool): whether the state is batched or not

    Returns:
```

albi3ro commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+            new_row_indices (str): Tensor indices of the state
+            **kwargs (dict): Stores indices calculated in `get_einsum_mapping`
+
+        Returns:
```
Mind adding a section like Args for the closure variables? I'm having a hard time figuring out which variables from the outer scope we are relying on.

```
Closure:
    val1
```

If we are not relying on any variables from outer scope, could we pull this out to be a global function? That can make things a lot easier to debug and follow.

Gabriel-Bottrill replied:
I am going to move it out. I will add all the expected kwargs to docstring too.

albi3ro commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+        if isinstance(measurementprocess, ExpectationMP):
+            # TODO add faster methods
+            # TODO add support for sparce Hamiltonians
+            if isinstance(measurementprocess.obs, Hamiltonian):
+                return sum_of_terms_method
+            if isinstance(measurementprocess.obs, Sum):
```
```suggestion
            if isinstance(measurementprocess.obs, (Hamiltonian, Sum)):
```

Gabriel-Bottrill replied:
Good catch thank you!

albi3ro commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
Do we happen to have any analytic formulas for expectation values/ probabilities with qutrit?  While it can be be difficult to get analytic formulas for anything more than the most basic cases, even those really basic examples can go a long way to instilling confidence in an implementation.

Not fully necessary, but usually nice to have.

Gabriel-Bottrill replied:
Yeah, that can be done. I will add one to the tests.

albi3ro reviewed (commented):
Looking in really nice shape :)

I left a couple questions and minor comments, but I'm pretty happy with this.

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+    for i, wire_to_trace in enumerate(wires_to_trace):
+        axis1 = wire_to_trace - i
+        axis2 = axis1 + num_state_wires - i
+
+        print(axis1, axis2)
+
+        state = math.trace(state, axis1=axis1, axis2=axis2)
```
```suggestion
        state = math.trace(state, axis1=axis1, axis2=axis2)
```

Gabriel-Bottrill replied:
I just realized I should find a better way to do this too, it is dependent on axis1 and axis2 that will not work for all interfaces. Einsum should work well here though.

Gabriel-Bottrill replied:
I added my new method, hopefully this still looks good

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+    """Finds the probabilities of measuring states in the computational basis
+
+    Args:
+        state (array[complex]): Input quantum state
+        num_wires (int): The number of wires the state represents
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+
+    Returns:
+        Tensorlike: Vector representing probability of each state
```
```suggestion
        state (array[complex]): input quantum state
        num_wires (int): the number of wires the state represents
        is_state_batched (bool): Boolean representing whether the state is batched or not

    Returns:
        Tensorlike: vector representing probability of each state
```
Minor formatting point for consistency.

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+    # take the real part so probabilities are not shown as complex numbers
+    # also reshape to flat
+    probs = math.reshape(math.real(probs), final_shape)
+    return math.where(probs < 0, -probs, probs)
```
Wait... `probs` should never be negative. Is this handling cases where they work out to very very small negative values due to precision issues? If so I think `math.clip` could be a better option

Gabriel-Bottrill replied:
Yes I agree, I can change that.

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+        state = apply_operation(op, state, is_state_batched=is_state_batched)
+
+    num_state_wires = get_num_wires(state, is_state_batched)
+    probs = _get_probs(state, num_state_wires, is_state_batched)
```
`_get_probs` is quite short and straightforward to follow, I think its contents could just be moved into here (though I see the rationale for defining it separately it, it doesn't look like it is being used anywhere else).

Gabriel-Bottrill replied:
I had it separately for its use in sample, but I end up using calculate probs instead anyway, I will move it out

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+    num_state_wires = get_num_wires(state, is_state_batched)
+    probs = _get_probs(state, num_state_wires, is_state_batched)
+
+    if mp_wires := measurementprocess.wires:
```
What is the `:=` doing?

Gabriel-Bottrill replied:
it defines `mp_wires = measurementprocess.wires` and does `if measurementprocess.wires` its used elsewhere in PennyLane.

albi3ro replied:
Walruses!

glassnotes replied:
Huh! Learned something new :)

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+    return math.real(trace)
+
+
+def reduce_density_matrix(  # TODO: ask if I should have state diagonalization gates?
```
```suggestion
def calculate_reduced_density_matrix(  # TODO: ask if I should have state diagonalization gates?
```

Gabriel-Bottrill replied:
I agree it's a better name.

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+    return math.einsum(einsum_indices, obs_mat, state)
+
+
+def trace_method(
```
```suggestion
def calculate_expval(
```

Recommend using a consistent naming scheme for the different measurement function option (I made a similar suggestion for `reduce_dm` as well). It also makes it more clear what it is doing here.

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+    Returns:
+        Tensorlike: the result of the measurement
+    """
+    return get_measurement_function(measurementprocess, state)(
```
What is the rationale for having `get_measurement_function` defined separately instead of having the same logic in here?

Gabriel-Bottrill replied:
To remain consistant with twhat was done with `default.qubit` and to make it separately callable in a test. It could be reformatted if you prefer.

albi3ro replied:
We did so in default qubit to better separate the logic for determining what method we want to use from the logic for those methods themselves. This allows better unit testing the choice of kernel without resorting to mockers and spying.

It's convenient for `default.qubit`, as we have four different kernels that we choose between based on a variety of constraints and optimizations. 

But that might premature optimization and speculative generality here.

Gabriel-Bottrill replied:
So would you like that to be removed?

glassnotes replied:
Let's leave it as-is, for ease of testing and consistency.

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,309 @@
+    )
+
+
+def sum_of_terms_method(
```
```suggestion
def calculate_expval_sum_of_terms(
```
(similar to other naming suggestions)

glassnotes commented on pennylane/devices/qutrit_mixed/utils.py:
```diff
@@ -33,7 +34,7 @@ def get_einsum_mapping(
 
     Returns:
-        str: indices mapping that defines the einsum
+        str: Indices mapping that defines the einsum
```
```suggestion
        str: indices mapping that defines the einsum
```
Keep consistent with rest of docs; no need for change

glassnotes commented on pennylane/devices/qutrit_mixed/utils.py:
```diff
@@ -68,6 +69,37 @@ def get_einsum_mapping(
 
 
+def resquare_state(state, num_wires):
+    """
+    Given a non-flat, potentially batched state, flatten it to a square matrix.
```
```suggestion
    """Given a non-flat, potentially batched state, flatten it to a square matrix.
```

glassnotes commented on pennylane/devices/qutrit_mixed/utils.py:
```diff
@@ -68,6 +69,37 @@ def get_einsum_mapping(
     )
 
 
+def resquare_state(state, num_wires):
```
Maybe rename as some variant of `reshape_state`? Since `square` also has the mathematical connotation of squaring.

Gabriel-Bottrill replied:
changed to `reshape_state_as_matrix`

glassnotes commented on pennylane/devices/qutrit_mixed/utils.py:
```diff
@@ -68,6 +69,37 @@ def get_einsum_mapping(
+    """Finds the number of wires associated with a state
+
+    Args:
+        state (TensorLike): A state that needs flattening
```
```suggestion
        state (TensorLike): A state that needs flattening
```
Update this docstring to reflect this method

Gabriel-Bottrill replied:
updated to better reflect method

glassnotes replied:
This docstring hasn't been updated; should reflect what the function is doing here

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -0,0 +1,507 @@
+class TestCurrentlyUnsupportedCases:
+    # pylint: disable=too-few-public-methods
+    def test_sample_based_observable(self, two_qutrit_state):
+        """Test sample-only measurements raise a notimplementedError."""
```
```suggestion
        """Test sample-only measurements raise a NotImplementedError."""
```

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -0,0 +1,507 @@
+        assert qml.math.allclose(res, expected)
+
+    def test_apply_observable_einsum_batched(self, obs, three_qutrit_batched_state, ml_framework):
+        """Tests that unbatched observables are applied correctly to a batched state."""
```
Does it make sense to test the other two cases (batched observable/unbatched state, and batched/batched)?

Gabriel-Bottrill replied:
Maybe, but I am not sure if observables are supposed to be batched I don't think they are for `default.qubit`. Currently the code won't work for batched observables, but it could be added by adding `...` in the `apply_observable_einsum` implementation.

glassnotes replied:
Okay if there is not batching support for the qubit case let's leave it out to make things consistent here, then.

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -0,0 +1,507 @@
+    @pytest.mark.parametrize(
+        "measurement, get_expected",
+        [
+            (qml.state(), lambda x: math.reshape(x, newshape=(9, 9))),
```
How come these are being expressed here as function calls, i.e., `qml.state()` instead of just `qml.state`?

Gabriel-Bottrill replied:
qml.state() returns a StateMP object, there is similar code in the `test_measure` file in `default.qubit`

glassnotes reviewed (changes requested):
Looking good @Gabriel-Bottrill ! Didn't look at the tests in as much detail as the rest of the code, but they looked fine at a glance.

Gabriel-Bottrill commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -0,0 +1,507 @@
+    def test_sample_based_observable(self, two_qutrit_state):
+        """Test sample-only measurements raise a notimplementedError."""
+        with pytest.raises(NotImplementedError):
+            _ = measure(qml.sample(wires=0), two_qutrit_state)
```
I agree II'll add that

Gabriel-Bottrill commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -0,0 +1,507 @@
+probs_processes = [qml.probs(wires=0), qml.probs(op=qml.GellMann(0, 8)), qml.probs(wires=[0, 1])]
+
+
+class TestNaNMeasurements:
```
I added them because they weren't to difficult to add and I saw it as an eassy place to add some interface testing, I can remove them if you like or keep them as is, whatever you see as best.

Gabriel-Bottrill commented:
changing `math.where(probs < 0, -probs, probs)` to `math.clip` broke tests for NaN values. I could remove the NaN value tests, but then I would probably need to add some testing for interfaces elsewhere. I would appreciate some input on what would be best.

Gabriel-Bottrill commented:
Updates since last reviews (Thursday):

- Fixed suggested issues:
   - doc-string consistency and capitalization
   - Names of methods
   - Added analytical tests for expval
   - Moved `_map_indices_apply_operation` out of `apply_observable_einsum`
   - ... (a few small changes I may have forgot)
- Removed NaN tests
- Made broadcasting tests test ml_interfaces


Concerns/questions:
- [ ] Should `get_measurement_function` be defined separately?
- [ ] Should I delete `measure_func=measure` in `calculate_expval_sum_of_terms`

Gabriel-Bottrill commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,301 @@
+    state_wire_indices = "".join(state_wire_indices_list)
+    final_state_wire_indices = "".join(final_state_wire_indices_list)
+
+    state = math.einsum(f"...{state_wire_indices}->...{final_state_wire_indices}", state)
```
I had to remove some einsums used for diagonalization and tracing because it breaks when trying to find the derivative. I did this in other places.

Here I left reduce dm as I assume it will be uncommon to take the derivative of a reduced density matrix (especially since it would have to be flattened to do so), I could fix it so that it works, but it would require that I copy a lot of code or abstract the base code to work with qutrits. I think it's best to leave as is for now

albi3ro replied:
Mind providing an example of it breaking derivatives? It seems like the ml frameworks should be really good at handling einsum, so I'm a little confused about what would be breaking.

Gabriel-Bottrill replied:
```
state = qml.numpy.array(state)

def f(x):
    new_state = x * state
    diag = qml.math.einsum("abcabc->abc", new_state)
    return qml.math.flatten(diag)

x = qml.numpy.array(0.2, requires_grad=True)

g = qml.jacobian(f)(x)
```

Returns error:
```
[/usr/local/lib/python3.10/dist-packages/numpy/core/einsumfunc.py](https://localhost:8080/#) in einsum(out, optimize, *operands, **kwargs)
   1369         if specified_out:
   1370             kwargs['out'] = out
-> 1371         return c_einsum(*operands, **kwargs)
   1372 
   1373     # Check the kwargs to avoid a more cryptic error later, without having to

ValueError: einstein sum subscripts string includes output subscript 'a' multiple times
```
The einsum subscripts are swapped during backpropogation so `"abcabc->abc"` => `"abc->abcabc"` can't have repeating values in RHS. The same error occurs with trace.

Gabriel-Bottrill reviewed (commented):
I had to remove some einsums used for diagonalization and tracing because it breaks when trying to find the derivative. I left reduce dm to be non differentiable.

mudit2812 commented on doc/releases/changelog-dev.md:
```diff
@@ -202,6 +210,7 @@
   are being raised unexpectedly.
   [(#5122)](https://github.com/PennyLaneAI/pennylane/pull/5122)
 
+
```
```suggestion
```
Just removing extra white space

mudit2812 commented on doc/releases/changelog-dev.md:
```diff
@@ -134,6 +134,7 @@
 * Improve the performance of circuit-cutting workloads with large numbers of generated tapes.
   [(#5005)](https://github.com/PennyLaneAI/pennylane/pull/5005)
 
+
```
```suggestion
```
Just removing unnecessary white space

mudit2812 commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,301 @@
+    is_state_batched: bool = False,
+    measure_func=measure,
+) -> TensorLike:
+    """Measure the expecation value of the state when the measured observable is a ``Hamiltonian`` or ``Sum``
```
```suggestion
    """Measure the expectation value of the state when the measured observable is a ``Hamiltonian`` or ``Sum``
```

mudit2812 commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,301 @@
+# pylint: disable=too-many-return-statements
+def get_measurement_function(
+    measurementprocess: MeasurementProcess,
+    state: TensorLike,  # pylint: disable=unused-argument
```
```suggestion
```

Is this just an artifact from copying over from `default.qubit`?

Gabriel-Bottrill replied:
Yes I think I left a comment about it, I was thinking I should remove it but kept it to check.

mudit2812 commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,301 @@
+    measurementprocess: ExpectationMP,
+    state: TensorLike,
+    is_state_batched: bool = False,
+    measure_func=measure,
```
I second removing this.

Gabriel-Bottrill replied:
I agree, removing it now

mudit2812 reviewed (commented):
Looks good to me! I don't have much to add other than the comments added by other reviewers. Happy to approve once those are resolved.

glassnotes commented on doc/releases/changelog-dev.md:
```diff
@@ -172,6 +172,13 @@
 * The module `pennylane/math/quantum.py` has now support for the min-entropy.
   [(#3959)](https://github.com/PennyLaneAI/pennylane/pull/3959/)
 
+* A function called `apply_operation` has been added to the new `qutrit_mixed` module found in `qml.devices` that applies operations to device-compatible states.
```
Sorry if I missed this in a previous discussion thread, but why is this showing up as a new change copied down here? It looks like it appears in two places in the file now, but the update for `measure` is only in this section.

Gabriel-Bottrill replied:
I think there was an error in merging I tried to fix it by deleting some copying I will revert this file to the error and fix it.

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,296 @@
+def _map_indices_apply_operation(
+    state_indices, row_indices, new_row_indices, **kwargs
+):  # pylint: disable=unused-argument
+    """Map indices to wires
+    Args:
```
```suggestion
    """Map indices to wires.
    
    Args:
```

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,296 @@
+
+
+def _map_indices_apply_operation(
+    state_indices, row_indices, new_row_indices, **kwargs
```
I noticed that this has a different signature than the `_map_indices_apply_channel` in `apply_operation.py`. The implementation here also doesn't make use of any `kwargs` at all. Recommend to make these consistent.

Gabriel-Bottrill replied:
Yes it is because it is called with a dictionary the kwargs is just to catch the extra args that aren't used, I replaced it so it just takes kwargs so that it looks the same as `_map_indices_apply_channel`. It will be in the next commit.

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,296 @@
+)
+from .apply_operation import apply_operation
+
+alphabet_array = math.asarray(list(alphabet))
```
Can just import this from `utils.py`

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,296 @@
+        is_state_batched (bool): whether the state is batched or not.
+
+    Returns:
+        TensorLike: the resulting resquared density_matrix.
```
```suggestion
        TensorLike: the resulting state or reduced density matrix.
```

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,296 @@
+        is_state_batched (bool): whether the state is batched or not.
+
+    Returns:
+        TensorLike: the result of the measurement.
```
Please make these more descriptive and tailored to the function (comment applies to other functions too)

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,296 @@
+
+    # if a probability is very small it may round to negative, undesirable.
+    # math.clip with None bounds breaks with tensorflow, using this instead:
+    probs = math.where(probs < 0, -probs, probs)
```
Recommend to just set it explicitly to 0 (if it is a small enough value so as to be negative, it is invalid anyways)

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -0,0 +1,296 @@
+    return get_measurement_function(measurementprocess)(measurementprocess, state, is_state_batched)
+
+
+def calculate_expval_sum_of_terms(
```
Please move definition of function to before the `get_measurement_function` so that it is not forward-referenced.

glassnotes commented on pennylane/devices/qutrit_mixed/utils.py:
```diff
@@ -68,6 +69,36 @@ def get_einsum_mapping(
+        num_wires (int): The number of wires the state represents
+
+    Returns:
+        Tensorlike: A squared state, with an extra batch dimension if necessary
```
```suggestion
        Tensorlike: A reshaped, square state, with an extra batch dimension if necessary
```

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -0,0 +1,466 @@
+        assert get_measurement_function(mp1) is calculate_reduced_density_matrix
+
+    def test_prod_calculate_expval_method(self):
+        """Test that the expectation value of a product uses the trace method."""
```
```suggestion
        """Test that the expectation value of a product uses the calculate expval method."""
```

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -0,0 +1,466 @@
+)
+@pytest.mark.parametrize("ml_framework", ml_frameworks_list)
+class TestApplyObservableEinsum:
+    """Tests that observables are applied correctly for trace method."""
```
```suggestion
    """Tests that observables are applied correctly for calculate expval method."""
```

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -0,0 +1,466 @@
+        ],
+    )
+    def test_hermitian_expval(self, observable, two_qutrit_state):
+        """Test that measurements of ternary hermitian work correctly."""
```
```suggestion
        """Test that measurements of qutrit hermitian work correctly."""
```

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -0,0 +1,466 @@
+        ),
+    ],
+)
+class TestExpValAnalytical:
```
Great! Thanks for working these cases out :tada:

glassnotes reviewed (changes requested):
Thanks for the updates, Gabe! Just some minor comments to address, otherwise looking good to me.

glassnotes commented on pennylane/devices/qutrit_mixed/measure.py:
```diff
@@ -39,31 +39,28 @@
-        row_indices (str): Indices that must be replaced with sums.
-        new_row_indices (str): Tensor indices of the state.
-        **kwargs (dict): Stores indices calculated in `get_einsum_mapping`.
+        **kwargs (dict): Stores indices calculated in `get_einsum_mapping`:
```
Thanks for updating this - as long as the other reviewers are okay with the convention here, this is fine with me.

glassnotes commented on tests/devices/qutrit_mixed/test_qutrit_mixed_measure.py:
```diff
@@ -58,6 +58,14 @@ def get_expval(op, state):
     return np.trace(op_mult_state)
 
 
+def test_probs_with_negative_on_diagonal():
```
Good call adding this test!

glassnotes reviewed (approved):
Thanks for making all the suggested changes, @Gabriel-Bottrill , all clear from my side!

mudit2812 reviewed (approved):
Looks good to me! Nice work @Gabriel-Bottrill !
